example/mailto/mailto.cpp

```diff
@@ -0,0 +1,773 @@
+    {
+        auto ps = u_.params();
+        auto it = ps.find(k);
+        if (it != ps.end())
+            return (*it).value;
+        return {};
+    }
+
+    friend mailto_rule_t;
+};
+
+/** Rule to match a mailto URL
+*/
+struct mailto_rule_t
+{
+    /// Value type returned by the rule
+    using value_type = mailto_view;
+
+    /// Parse a sequence of characters into a mailto_view
+    urls::result< value_type >
+    parse( char const*& it, char const* end ) const noexcept
+    {
+        urls::result<urls::url_view> r =
+            urls::grammar::parse(it, end, urls::absolute_uri_rule);
+        if (!r)
+            return r.error();
+
+        mailto_view m;
+        m.u_ = *r;
+        if (m.u_.scheme() != "mailto")
+            return urls::grammar::error::invalid;
+        if (m.u_.has_authority())
+            return urls::grammar::error::invalid;
+        if (!is_to(m.u_.encoded_path()))
+            return urls::grammar::error::invalid;
+        for (auto p: m.u_.encoded_params())
+            if (!is_header(p))
+                return urls::grammar::error::invalid;
+
+        return m;
+    }
+
+private:
+    // A long set of functions to validate email
+    // addresses according to RFC6068
+    static
+    bool
+    is_to(urls::string_view s) noexcept
+    {
+        if (s.empty())
+            return true;
+        auto p = s.find(',');
+        while (p != urls::string_view::npos)
+        {
+            if (!is_address(s.substr(0, p)))
+                return false;
+            s = s.substr(p + 1);
+            p = s.find(',');
+        }
+        if (!is_address(s))
+            return false;
+        return true;
+    }
+
+    static
+    bool
+    is_address(urls::string_view s) noexcept
+    {
+        // addr-spec = local-part "@" domain
+        std::size_t p = s.find('@');
+        if (p == urls::string_view::npos)
+            return false;
+        return is_local_part(s.substr(0, p)) &&
+               is_domain(s.substr(p + 1));
+    }
+
+    static
+    bool
+    is_local_part(urls::string_view s) noexcept
+    {
+        // local-part = dot-atom-text / quoted-string
+        return is_dot_atom_text(s) ||
+               is_quoted_string(s);
+    }
+
+    static
+    bool
+    is_dot_atom_text(urls::string_view s) noexcept
+    {
+        // dot-atom-text = 1*atext *("." 1*atext)
+        if (s.empty())
+            return false;
+        bool prev_dot = true;
+        for (std::size_t i = 0; i < s.size(); ++i)
+        {
+            if (prev_dot)
+            {
+                if (is_atext(s[i]))
+                    prev_dot = false;
+                else
+                    return false;
+            }
+            else if (!is_atext(s[i]))
+            {
+                if (s[i] == '.')
+                    prev_dot = true;
+                else
+                    return false;
+            }
+        }
+        return true;
+    }
+
+    static
+    bool
+    is_atext(char c) noexcept
```

Comment:
should this be expressed as a _CharSet_ ?

Reply:
Technically, all these private functions could be expressed as rules and charsets. I just thought that would be very long but I can experiment with it.

Reply:
I would hope it would be comparable, but even if it is longer, if we are parsing something we should use the idioms offered by the library...